File: libs/fleet/sdk-bindings/typescript/test/public-api.ts
import {
  connect,
  type Claim,
  type CreateClaimRequest,
  type CreatePoolRequest,
  type Configuration,
  type Pool,
  type PoolSpec,
  type Sandbox,
  type ServiceFetch,
} from "../src/index.js"

const configuration: Configuration = {
  baseUrl: "https://run.example",
  oauth: { tokenUrl: "https://auth.example/token", clientId: "id", clientSecret: "secret" },
}
const poolSpec: PoolSpec = {
  replicas: 1,
  template: { containerDiskImage: "example/workspace:latest", cpuCores: 4, memory: "4Gi" },
}
const pool: Pool = {
  apiVersion: "cua.ai/v1",
  kind: "OSGymWorkspacePool",
  metadata: { namespace: "pool-one", name: "pool-one" },
  spec: { replicas: 1, template: { containerDiskImage: "example/workspace:latest", cpuCores: 4, memory: "4Gi" } },
}
const claim: Claim = {
  apiVersion: "osgym.cua.ai/v1alpha1",
  kind: "OSGymSandboxClaim",
  metadata: { namespace: "pool-one", name: "claim-one" },
  spec: { sandboxTemplateRef: { name: "pool-one-template" } },
}

async function contract(): Promise<Sandbox> {
  const sdk = await connect(configuration)
  const poolRequest: CreatePoolRequest = { namespace: "pool-one", spec: poolSpec }
  const createdPool = await sdk.createPool(poolRequest)
  await sdk.listPools(createdPool.metadata.namespace)
  await sdk.getPool(pool)
  await sdk.updatePool(pool)
  const claimRequest: CreateClaimRequest = { pool: createdPool }
  const createdClaim = await sdk.createClaim(claimRequest)
  await sdk.listClaims(createdClaim.metadata.namespace)
  await sdk.getClaim(claim)
  await sdk.updateClaim(claim)
  const sandbox = await sdk.waitClaim(claim)
  const service: ServiceFetch = sdk.serviceFetch(sandbox, "mcp")
  const response: Response = await service("/mcp", { method: "POST", body: "{}" })
  void response
  // @ts-expect-error URL-based authenticated clients are intentionally not public.
  void sdk.rawClient
  await sdk.deleteClaim(claim)
  await sdk.deletePool(pool)
  sdk.close()
  return sandbox
}
void contract

File: libs/fleet/sdk-bindings/rust/tests/public_api.rs
use cyclops_sdk_host::{
    Claim, Configuration, CreateClaimRequest, CreatePoolRequest, Pool, Sandbox, Sdk, SdkError,
    ServiceHttpResponse,
};

#[test]
fn typed_resource_api_only() {
    let _: fn(Configuration) -> Result<Sdk, String> = Sdk::connect;
    let _: fn(&mut Sdk, CreatePoolRequest) -> Result<Pool, String> = Sdk::create_pool;
    let _: fn(&mut Sdk, &str) -> Result<Vec<Pool>, String> = Sdk::list_pools;
    let _: fn(&mut Sdk, Pool) -> Result<Pool, String> = Sdk::get_pool;
    let _: fn(&mut Sdk, Pool) -> Result<Pool, String> = Sdk::update_pool;
    let _: fn(&mut Sdk, Pool) -> Result<(), String> = Sdk::delete_pool;
    let _: fn(&mut Sdk, CreateClaimRequest) -> Result<Claim, String> = Sdk::create_claim;
    let _: fn(&mut Sdk, &str) -> Result<Vec<Claim>, String> = Sdk::list_claims;
    let _: fn(&mut Sdk, Claim) -> Result<Claim, String> = Sdk::get_claim;
    let _: fn(&mut Sdk, Claim) -> Result<Claim, String> = Sdk::update_claim;
    let _: fn(&mut Sdk, Claim) -> Result<(), String> = Sdk::delete_claim;
    let _: fn(&mut Sdk, Claim) -> Result<cyclops_sdk_host::Sandbox, String> = Sdk::wait_claim;
}

fn service_http_contract(sdk: &mut Sdk, sandbox: &Sandbox) -> Result<(), SdkError> {
    let mut service = sdk.service_client(sandbox, "mcp")?;
    let _: ServiceHttpResponse = service
        .post("/mcp")
        .header("content-type", "application/json")
        .body("{}")
        .send()?;
    Ok(())
}

#[test]
fn service_http_is_sdk_owned() {
    let _: fn(&mut Sdk, &Sandbox) -> Result<(), SdkError> = service_http_contract;
}
